book/src/04_3_i2c.md

```diff
@@ -2,16 +2,99 @@
 
 ## Introduction
 
-TODO describe protocol & purpose
-TODO https://learn.sparkfun.com/tutorials/i2c
+The I2C (Inter-Integrated Circuit) protocol allows multiple peripheral chips to communicate with one or more controller chips. Many devices can be connected to the same I2C bus and messages can be send to a particular device by specifying its I2C address. The protocol requires two signal wires and can only be used for short distance communications within one device. 
 
+One of the signal lines is for data (SDA) and the other is for the clock signal (SCL). An I2C transaction consists of one or more messages. Each message begins with a start symbol. A message is either write or read, indicated by the next bit in the row. Following this comes the actual message in form of bytes. The message ends with a stop symbol. The clock signal rises and falls in a specified frequency.
```

Comment:
```suggestion
One of the signal lines is for data (SDA) and the other is for the clock signal (SCL). The lines are pulled-high by default with some resistors fitted somewhere on the bus. Any device on the bus (or even multiple devices simultaneously) can 'pull' either or both of the signal lines low. This means that no damage occurs if two devices try and talk on the bus at the same time - the messages are merely corrupted (and detectably so).

An I²C _transaction_ consists of one or more _messages_. Each _message_ is comprised of a _start symbol_, some _words_, and finally a _stop symbol_ (or another _start symbol_ if there is a follow-on message). Each word is eight bits, followed by an ACK (0) or NACK (1) bit which is sent by the recipient to indicate whether the word was received and understood correctly. The first word indicates both the 7-bit address of the device the message is intended for, plus a bit to indicate if the device is being _read from_ or being _written to_. If there is no device of that address on the bus, the first word will naturally have a 'NACK' after it (because there is no device driving the SDA line low to generate an 'ACK' bit) and so you know there is no device present.

The clock frequency of the SCL line is usually 100 kHz but sometimes faster speeds like 400 kHz (or higher) are supported.

To read three bytes from an EEPROM device, the sequence will be something like:

| Step | Controller Sends         | Peripheral Sends                  |
| ---- | ------------------------ | --------------------------------- |
| 1.   | START                    |                                   |
| 2.   | Device Address + W       |                                   |
| 3.   |                          | ACK                               |
| 4    | High EEPROM Address byte |                                   |
| 5    |                          | ACK                               |
| 6    | Low EEPROM Address byte  |                                   |
| 7    |                          | ACK                               |
| 8    | START                    |                                   |
| 9    | Device Address + R       |                                   |
| 10   |                          | ACK                               |
| 11   |                          | Data Byte from EEPROM Address     |
| 12   | ACK                      |                                   |
| 13   |                          | Data Byte from EEPROM Address + 1 |
| 14   | ACK                      |                                   |
| 15   |                          | Data Byte from EEPROM Address + 2 |
| 16   | NAK (i.e. end-of-read)   |                                   |
| 17   | STOP                     |                                   |
```